### PR TITLE
Add edge test cases for 'convert::from_string'

### LINKED ISSUE #2055: 'convert' is broken for edge cases
## Required information

**Operating system:**
all

**Compiler version:**
all

**Eclipse iceoryx version:**
all

**Observed result or behaviour:**
Conversions from strings like `nan`, `inf` or `DBL_MAX` for floating point types or `ULLONG_MAX` for uint64_t fail.

**Expected result or behaviour:**
All valid values should be converted.

**Conditions where it occurred / Performed steps:**
Try to convert `inf` to `double` or `18446744073709551615` to `uint64_t`

## Detailed description of the problem

The conversion functions have some weird way to tell the user something went wrong. Let's take `strtoull` for example
```c
unsigned long long int strtoull(const char *nptr, char **endptr,  int base);
```
There are three scenarios:
1. `nptr == endptr`, the conversion failed because there were no valid characters
2.  `nptr != endptr && *endptr != '\0'`, the conversion was (probably) successful and it depends on the user if this is correct because not all characters were converted, e.g. if the user is passed `123ab` then ` *endptr` would be `a` but if the user expected the whole string to be a number, this should be an error
3. `nptr != endptr && *endptr != '\0'` , the conversion was (probably) successful

2 and 3 might still have failed if the return value was `ULLONG_MAX` **and** the `errno` is set to `ERANGE`.

All of this has to be taken care of or the conversion will fail on edge cases

## Test program

This small test program can be used to observe the behavior for string to integer and string to floating point conversion.
```cpp
#include <cstdlib>
#include <iostream>

int main() {
    struct Test {
        const char* name;
        const char* value;
    };

    const Test integer_tests[] {{"ZERO", "0"},
                                {"ULLONG_MAX", "18446744073709551615"},
                                {"ULLONG_MAX + 1", "18446744073709551616"},
                                {"ULLONG_MAX + ascii", "18446744073709551615q"},
                                {"ULLONG_MAX + 1 + ascii", "18446744073709551616q"},
                                {"String 'COFFEE'", "COFFEE"}
                               };

    std::cout << "#### Integer" << std::endl;
    for(const auto& test: integer_tests) {
        std::cout << "==== " << test.name << std::endl;
        char* end{nullptr};
        errno = 0;
        auto retval = strtoull(test.value, &end, 10);
        std::cout << "tst val: " << test.value << std::endl;
        std::cout << "ret val: " << retval << std::endl;
        std::cout << "errno: " << errno << std::endl;
        std::cout << "val ptr: "<< std::hex << static_cast<const void*>(test.value) << std::dec << std::endl;
        std::cout << "end ptr: " << std::hex << static_cast<void*>(end) << std::dec << " | " << static_cast<uint16_t>(*end) << " | " << end << std::endl;
        std::cout << std::endl;
    }


    const Test floating_point_tests[] {{"PI", "3.14"},
                                       {"DBL_MAX", "1.79769e+308"},
                                       {"nan", "nan"},
                                       {"inf", "inf"},
                                       {"Too large", "2e411234535345"},
                                       {"String 'COFFEE'", "COFFEE"}
                                      };

    std::cout << "#### Floating point" << std::endl;
    for(const auto& test: floating_point_tests) {
        std::cout << "==== " << test.name << std::endl;
        char* end{nullptr};
        errno = 0;
        auto retval = strtod(test.value, &end);
        std::cout << "tst val: " << test.value << std::endl;
        std::cout << "ret val: " << retval << std::endl;
        std::cout << "errno: " << errno << std::endl;
        std::cout << "val ptr: "<< std::hex << static_cast<const void*>(test.value) << std::dec << std::endl;
        std::cout << "end ptr: " << std::hex << static_cast<void*>(end) << std::dec << " | " << static_cast<uint16_t>(*end) << " | " << end << std::endl;
        std::cout << std::endl;
    }
}
```

## Tasks

- [x] refactor code and make current tests pass
- [x] additional tests for edge cases
- [x] determine whether the subnormal float behavior should be unified; currently MSVC converts subnormal floats but GCC and clang does not

### PR DESCRIPTION
**Pull request by Dennis40816**
## Pre-Review Checklist for the PR Author

1. [x] Add a second reviewer for complex new features or larger refactorings
1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] All touched (C/C++) source code files from `iceoryx_hoofs` are added to `./clang-tidy-diff-scans.txt`
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Notes for Reviewer

Note: The items in **bold** indicate that they have already been implemented.

### Signed Integral Types

1. **`signed char`**
    - **`CHAR_MIN`** (-128)
    - **`CHAR_MIN - 1`** (-129)
    - **`CHAR_MAX`** (127)
    - **`CHAR_MAX + 1`** (128)
2. **`signed short` [tested]**
    - **`SHRT_MIN`** (-32768)
    - **`SHRT_MIN - 1`** (-32769)
    - **`SHRT_MAX`** (32767)
    - **`SHRT_MAX + 1`** (32768)
3. **`signed int` [tested]**
    - **`INT_MIN`** (-2147483648)
    - **`INT_MIN - 1`** (-2147483649)
    - **`INT_MAX`** (2147483647)
    - **`INT_MAX + 1`** (2147483648)
4. **`signed long`**
    - **`LONG_MIN`** (-2147483648 / -9223372036854775808)
    - **`LONG_MIN - 1`** (-2147483649 / -9223372036854775809)
    - **`LONG_MAX`** (2147483647 / 9223372036854775807)
    - **`LONG_MAX + 1`** (2147483648 / 9223372036854775808)
5. **`signed long long`**
    - **`LLONG_MIN`** (-9223372036854775808)
    - **`LLONG_MIN - 1`** (-9223372036854775809)
    - **`LLONG_MAX`** (9223372036854775807)
    - **`LLONG_MAX + 1`** (9223372036854775808)

### Unsigned Integral Types

1. **`unsigned char`**
    - **`0`**
    - **`-1`**
    - **`UCHAR_MAX`** (255)
    - **`UCHAR_MAX + 1`** (256)
2. **`unsigned short` [tested]**
    - **`0`**
    - **`-1`**
    - **`USHRT_MAX`** (65535)
    - **`USHRT_MAX + 1`** (65536)
3. **`unsigned int` [tested]**
    - **`0`**
    - **`-1`**
    - **`UINT_MAX`** (4294967295)
    - **`UINT_MAX + 1`** (4294967296)
4. **`unsigned long`**
    - **`0`**
    - **`-1`**
    - **`ULONG_MAX`** (4294967295 / 18446744073709551615)
    - **`ULONG_MAX + 1`** (4294967296 / 18446744073709551616)
5. **`unsigned long long`**
    - **`0`**
    - **`-1`**
    - **`ULLONG_MAX`** (18446744073709551615)
    - **`ULLONG_MAX + 1`** (18446744073709551616)

### Floating Point Type

1. **`float`**
    - **`FLT_MIN`** (using `numeric_limits`)
    - **`FLT_MIN - epsilon`** (using `numeric_limits`) -> should failed
    - **`std::numeric_limits<float>::lowest()`** (using `numeric_limits`)
    - `std::numeric_limits<float>::lowest() - epsilon` (Not Impl)
    - **`FLT_MAX`** (using `numeric_limits`)
    - `FLT_MAX + epsilon` (Not Impl)
    - **`NAN`**
    - **`INF`**
    - **`-INF`**
2. **`double`**
    - **`DBL_MIN`** (using `numeric_limits`)
    - **`DBL_MIN - epsilon`** (using `numeric_limits`) -> should failed
    - **`std::numeric_limits<double>::lowest()`** (using `numeric_limits`)
    - `std::numeric_limits<double>::lowest() - epsilon` (Not Impl)
    - **`DBL_MAX`** (using `numeric_limits`)
    - `DBL_MAX + epsilon` (Not Impl)
    - **`NAN`**
    - **`INF`**
    - **`-INF`**
3. **`long double`**
    - **`LDBL_MIN`** (using `numeric_limits`)
    - **`LDBL_MIN - epsilon`** (using `numeric_limits`) -> should failed
    - **`std::numeric_limits<long double>::lowest()`** (using `numeric_limits`)
    - `std::numeric_limits<long double>::lowest() - epsilon` (Not Impl)
    - **`LDBL_MAX`** (using `numeric_limits`)
    - `LDBL_MAX + epsilon` (Not Impl)
    - **`NAN`**
    - **`INF`**
    - **`-INF`**

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] All touched (C/C++) source code files from `iceoryx_hoofs` have been added to `./clang-tidy-diff-scans.txt`
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Closes #2055

### CODE REVIEW

**Comment by Dennis40816** [comment]
@elBoberido I think it's ready to be reviewed :) 

Btw, the following program has different behavior between MSVC and MinGW (align with Linux).

```cpp
#include <iostream>
#include <cmath>
#include <string>

int main ()
{
    std::string sub_normal_float = "1e-45"; // a subnormal float number
    auto ret = strtof(sub_normal_float.c_str(), nullptr);
    std::cout << "ret: " << ret << ", errno: " << errno << std::endl;
    std::cout << "Is normal: " << std::isnormal(ret);
}
```

- MSVC

```bash
# cl test.cpp -o test_win
ret: 1.4013e-45, errno: 0
Is normal: 0
```
- MinGW

```bash
# g++ test.cpp -o test_g++
ret: 1.4013e-45, errno: 34
Is normal: 0
```

However, I have been unable to locate a MSVC document that discusses this specific issue (maybe I used wrong key words). If you have any relevant sources, sharing them would be greatly appreciated.

**Comment by elBoberido** [comment]
@Dennis40816 I might not have time to have a closer look today but I think I can do the review tomorrow. Just a small thing. I noticed that there are multiple tests per test case. This is not necessarily bad and we also do not need to have one test per test case but what would be nice is if the tests cases are split between edge cases for valid conversion and the ones for invalid conversions, e.g. 0 and 255 for uint8_t in a `fromString_UnsignedChar_EdgeCase_InRange_Success` test case and -1 and 256 for uint8_t in a `fromString_UnsignedChar_EdgeCase_OufOfRange_Fail` test case. It should have been like that in the first place to better distinguish the between tests for valid strings and invalid ones.

Regarding the issue with the floating point. I'm also not sure about this behavior. I also could not find anything and the only resource I know which might bring some linght into the dark is https://dl.acm.org/doi/pdf/10.1145/103162.103163.

File: iceoryx_hoofs/test/moduletests/test_utility_convert.cpp [hunk 1]
```diff
@@ -303,59 +312,509 @@ TEST_F(convert_test, fromString_LongInt_Fail)
...
+    {
+        auto inf_ret = iox::convert::from_string<long double>(v.c_str());
+        ASSERT_THAT(inf_ret.has_value(), Eq(true));
+        ASSERT_THAT(std::isinf(inf_ret.value()), Eq(true));
+    }
+}
+
+/// SPECIAL FLOATING POINT TYPE EDGE CASES END
+
 TEST_F(convert_test, fromString_cxxString)
```
Comment (elBoberido) [inline iceoryx_hoofs/test/moduletests/test_utility_convert.cpp:818]: Could you please rename this to `fromString_ioxString`?

**Review by elBoberido — COMMENTED** [review]
Looks good overall :)

Could you add a few more values to the floating point in range success test cases
- `0`
- `0.`
- `0.0`
- `.0`
This could be done in a loop similar to NAN and INF.

Maybe also another one with just `.1`.

For the integer there should be a test with a trailing char, e.g. `42a` and/or `73 ` which should fail.

**Review by elBoberido — APPROVED** [review]
Good job :)

Btw, next Tuesday is the next developer meetup and the week after there is another one to introduce the Rust based iceoryx2. You are welcome to join.

**Comment by Dennis40816** [comment]
I'm glad to hear that. Additionally, I'm excited to learn about the rust iceoryx2 project. In fact, it popped up in my GitHub recommendations a few days ago :) If I get the chance, I'll be there.

**Comment by Dennis40816** [comment]
Also, I can't wait to start working on the next issue. I remember you mentioned a refactoring plan related to serialization before. I think we might be able to start from there.

**Comment by elBoberido** [comment]
I think I was a little bit too enthusiastic in merging the PR. There was still the open question whether the behavior on subnormal floats should be unified. I'll re-open the issue and add this question as a task.

Unfortunately it seems there is another issues which showed up only after the merge to master. The CI on aarch64 fails for one edge case test. This is the error
```
/tmp/cirrus-ci-build/iceoryx_hoofs/test/moduletests/test_utility_convert.cpp:687: Failure
Value of: double_min.has_value()
Expected: is equal to true
  Actual: false (of type bool)
#### Log start ####
2024-01-07 18:23:11.471 [Debug]: ERANGE triggered during conversion of string: '2.22507385850720138e-308'
#### Log end ####
```
I'm not quite sure why it happens. Maybe an update on the docker container the CI runs. I had a similar issue where the Bazel build broke after a merge without having touched anything and it turned out that Bazel got updated to v7.0.

Could you look into this issue? A quick workaround could be to disable the test preliminary and buy time to fix it without pressure.

**Comment by Dennis40816** [comment]
> I think I was a little bit too enthusiastic in merging the PR. There was still the open question whether the behavior on subnormal floats should be unified. I'll re-open the issue and add this question as a task.
> 

Agreed. I personally believe that the convert behavior in iox should maintain cross-platform consistency.
> Unfortunately it seems there is another issues which showed up only after the merge to master. The CI on aarch64 fails for one edge case test. This is the error
> 
> ```
> /tmp/cirrus-ci-build/iceoryx_hoofs/test/moduletests/test_utility_convert.cpp:687: Failure
> Value of: double_min.has_value()
> Expected: is equal to true
>   Actual: false (of type bool)
> #### Log start ####
> 2024-01-07 18:23:11.471 [Debug]: ERANGE triggered during conversion of string: '2.22507385850720138e-308'
> #### Log end ####
> ```
> 
> I'm not quite sure why it happens. Maybe an update on the docker container the CI runs. I had a similar issue where the Bazel build broke after a merge without having touched anything and it turned out that Bazel got updated to v7.0.
> 
> Could you look into this issue? A quick workaround could be to disable the test preliminary and buy time to fix it without pressure.

Okay. I will disable this test first. Also, could you provide the link which action failed? I can't find the specific action.

**Comment by elBoberido** [comment]
> > I think I was a little bit too enthusiastic in merging the PR. There was still the open question whether the behavior on subnormal floats should be unified. I'll re-open the issue and add this question as a task.
> 
> Agreed. I personally believe that the convert behavior in iox should maintain cross-platform consistency.

Then let's just do it an return an ERANGE for subnormal floats :)

**Comment by Dennis40816** [comment]
Oh

> > > I think I was a little bit too enthusiastic in merging the PR. There was still the open question whether the behavior on subnormal floats should be unified. I'll re-open the issue and add this question as a task.
> > 
> > 
> > Agreed. I personally believe that the convert behavior in iox should maintain cross-platform consistency.
> 
> Then let's just do it an return an ERANGE for subnormal floats :)

Then, it is tentatively scheduled to be resolved in #2155.
